lib/shopify_api/rest.ex

```diff
@@ -11,25 +11,67 @@ defmodule ShopifyAPI.REST do
   alias ShopifyAPI.JSONSerializer
   alias ShopifyAPI.REST.Request
 
-  @doc false
-  def get(%AuthToken{} = auth, path, params \\ []) do
-    Request.perform(auth, :get, path, "", params)
+  @doc """
+  Underlying utility retrieval function. The options passed affect both the
+  return value and, ultimately, the number of requests made to Shopify.
+  Options:
+    * `:pagination` - Can be `:none`, `:stream`, or `:auto`. Defaults to :auto
+      `:none` will only return the first page. You won't have access to the headers to manually
+      paginate.
+      `:auto` will block until all the pages have been retrieved and concatenated together.
+      `:stream` will return a `Stream`, prepopulated with the first page.
+  """
+  @spec get(AuthToken.t(), path :: String.t(), keyword(), keyword()) ::
+          {:ok, %{required(String.t()) => [map()]}} | Enumerable.t()
+  def get(%AuthToken{} = auth, path, params \\ [], options \\ []) do
+    case pagination(options) do
+      :none ->
+        with {:ok, response} <- Request.perform(auth, :get, path, "", params) do
+          {:ok, fetch_body(response)}
+        end
+
+      :stream ->
+        Request.stream(auth, path, params)
+
+      _auto_or_nil ->
+        auth
+        |> Request.stream(path, params)
+        |> Enum.to_list()
+    end
   end
 
   @doc false
   def post(%AuthToken{} = auth, path, object \\ %{}) do
-    with {:ok, body} <- JSONSerializer.encode(object),
-         do: Request.perform(auth, :post, path, body)
+    with {:ok, body} <- JSONSerializer.encode(object) do
+      perform_request(auth, :post, path, body)
+    end
   end
 
   @doc false
   def put(%AuthToken{} = auth, path, object) do
-    with {:ok, body} <- JSONSerializer.encode(object),
-         do: Request.perform(auth, :put, path, body)
+    with {:ok, body} <- JSONSerializer.encode(object) do
+      perform_request(auth, :put, path, body)
+    end
   end
 
   @doc false
-  def delete(%AuthToken{} = auth, path) do
-    Request.perform(auth, :delete, path)
+  def delete(%AuthToken{} = auth, path), do: perform_request(auth, :delete, path)
+
+  defp perform_request(auth, method, path, body \\ "") do
+    with {:ok, response} <- Request.perform(auth, method, path, body),
+         response_body <- fetch_body(response) do
+      {:ok, response_body}
+    end
+  end
+
+  defp fetch_body(http_response) do
+    with {:ok, map_fetched} <- Map.fetch(http_response, :body),
+         {:ok, body} <- map_fetched,
+         do: body
+  end
+
+  @spec pagination(keyword) :: atom | nil
+  defp pagination(options) do
+    Keyword.get(options, :pagination, Application.get_env(:shopify_api, :pagination, :none))
```

Comment:
Should this default be `:auto` instead of `:none`, as indicated in the doc comments above?

Reply:
Yes, good catch. I had been finding it easier to test this ^ way but I'll make sure to change it back.